content/blog/how-to-store-tokens-safely/index.md

```diff
@@ -0,0 +1,59 @@
+---
+title: How to Store Tokens Safely
+date: "2019-10-29"
+description: "Let's talk about how to store your tokens safely using cookies and lambdas."
+---
+
+How to store tokens safely? It's a common question on frontend community and there is some alternatives to this. In this post, I'll show you how to do this using cookies and lambdas.
+
+Before we get started, we need to know use local storage is not a good idea, as malicious scripts can iterate over the local storage and send the info to an external API. So if you have any kind of sensitive data there, you're screwed. We also need to know the easiest way of storing a token safely is just keeping it in memory, but this option has some pitfalls in UX. So what should we do?
+
+Cookies are a good tool for this scenario, but of course you should follow some guidelines to make it safe such as set the cookie as `secure`, `httpOnly` and `SameSite=strict`. The `secure` flag ensures that the cookie is sent only over `https` connections, the `httpOnly` makes the cookie inaccessible on client side so malicious scripts cannot see your sensitive data in `document.cookies` and `SameSite=strict` blocks cookies from being sent to an external domain.
+
+Since the cookie is not available on client side, we need to create some backend logic to fetch the token and store in memory. You can create a BFF(backend for frontends) using a web framework like Rails, Django, Phoenix, Laravel, etc. or you can use lambdas which are very easy and fast to move on. I recommend that you use Netlify Functions or Zeit. Let's see how it looks like.
```

Comment:
> Later in the article, it's evident that the frontend code (the app) is storing the token in memory. If that's the intention, the memory storage in this paragraph needs to be corrected. Also, I would strongly suggest that it is not the app storing the token, but the browser.

Both assumptions are right!

> If httpOnly restricts access from client code, then the frontend code cannot access it at all, correct? When given a cookie by the backend, the browser stores the cookie and includes it in subsequent requests. If this description matches your understanding, then I'd also suggest a different description than "store it in memory". Browsers operationally keep cookies in memory, but they also persist cookies in local file storage.

Yeah, makes sense. The intention here when I said "memory" is store it in a run time variable instead of put it on local storage, session storage or cookies.